src/tests/Microsoft.PowerFx.Interpreter.Tests/RecalcEngineAsyncTests.cs

```diff
@@ -0,0 +1,221 @@
+            Assert.Equal(12.0, result.ToObject());
+        }
+
+        private class Helper
+        {
+            public TaskCompletionSource<FormulaValue> _waiter = new TaskCompletionSource<FormulaValue>();
+
+            public async Task<FormulaValue> Worker2(FormulaValue[] args, CancellationToken cancel)
+            {
+                await Task.Yield();
+                var result = await _waiter.Task;
+
+                var n = ((NumberValue)result).Value;
+                var x = FormulaValue.New(n * 2);
+
+                return x;
+            }
+        }
+
+        // Verify a custom function can return a non-completed task. 
+        [Fact]
+        public async Task VerifyFunctionIsAsync()
+        {
+            var helper = new Helper();
+            var func = new CustomAsyncTexlFunction("CustomAsync", DType.Number, DType.Number)
+            {
+                _impl = helper.Worker2
+            };
+
+            var config = new PowerFxConfig(null);
+            config.AddFunction(func);
+
+            var engine = new RecalcEngine(config);
+
+            // Can be invoked. 
+            var cts = new CancellationTokenSource();
+
+            var task = engine.EvalAsync("CustomAsync(3)", cts.Token);
+            await Task.Yield();
+
+            // custom func is blocking on our waiter
+            await Task.Delay(TimeSpan.FromMilliseconds(5)); 
+            Assert.False(task.IsCompleted);
+
+            helper._waiter.SetResult(FormulaValue.New(15));
+
+            var result = await task;
+
+            Assert.Equal(30.0, result.ToObject());
+        }
+        
+        private static async Task<FormulaValue> WorkerWaitForCancel(FormulaValue[] args, CancellationToken cancel)
+        {
+            // Block forever until cancellation. 
+            await Task.Delay(-1, cancel); // throws TaskCanceledException
+
+            throw new InvalidOperationException($"Shouldn't get here");
+        }
+
+        // Verify cancellation 
+        [Fact]
+        public async Task Cancel()
+        {
+            var func = new CustomAsyncTexlFunction("CustomAsync", DType.Number, DType.Number)
+            {
+                _impl = WorkerWaitForCancel
+            };
+
+            var config = new PowerFxConfig(null);
+            config.AddFunction(func);
+
+            var engine = new RecalcEngine(config);
+
+            // Can be invoked. 
+            var cts = new CancellationTokenSource();
+
+            var task = engine.EvalAsync("CustomAsync(3)", cts.Token);
+            await Task.Yield();
+
+            // custom func is blocking on our waiter
+            await Task.Delay(TimeSpan.FromMilliseconds(5));
+            Assert.False(task.IsCompleted);
+
+            cts.Cancel();
+
+            await Assert.ThrowsAsync<TaskCanceledException>(async () => { await task; });
+        }
+
+        // Verify cancellation 
+        [Fact]
+        public async Task InfiniteLoop()
```

Comment:
>InfiniteLoop [](http://example.com/codeflow?start=26&length=12)

Even for purely sync functions, we still need a way to abort a long running function.